### PR TITLE
Make dependencies, max-version/min-version mandatory

### LINKED ISSUE #33241: Disallow enabling apps without min-version requirement
Remember this:
image

It's time to implement it in OC 11.

@DeepDiver1975 @VicDeo

### PR DESCRIPTION
**Pull request by VicDeo**
## Description
- [x]  remove deprecated **require**, **requiremin**, **requiremax** '' fileds support
- [x] make **dependencies** field mandatory
- [x] parse min/max from  **dependencies** only, treat missing min/max as an error

## Related Issue
- Closes https://github.com/owncloud/core/issues/33241

## Screenshot
![blockedapp](https://user-images.githubusercontent.com/991300/48080399-986ca580-e1fe-11e8-9891-9791a1eb97d9.png)


## How Has This Been Tested?
Unit tests

```
> php occ app:enable files_antivirus
files_antivirus enabled

> php occ app:enable user_persona  

In app.php line 1129:                                                                                                                           
  App "Mozilla Persona user backend" cannot be installed because the following dependencies are not fulfilled: missing minimal ownCl  
  oud version in appinfo.xml.                                                                                                         
  missing maximal ownCloud version in appinfo.xml.                                                                                    
                                                                                                                                      
app:enable [-g|--groups GROUPS] [--] <app-id>

```

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Technical debt
- [ ] Tests

## Checklist:
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:
<!-- In case of incomplete PR, please list the open tasks here -->
- [ ] Backport (if applicable set "backport-request" label and remove when the backport was done)

### CODE REVIEW

File: lib/private/App/CodeChecker/InfoChecker.php [hunk 1]
```diff
@@ -65,6 +65,12 @@ class InfoChecker extends BasicEmitter {
 		'standalone',
 	];
 
+	private $mandatoryAttributes = [
+		'dependencies' => [
+			'min-version'
```
Comment (PVince81) [inline lib/private/App/CodeChecker/InfoChecker.php]: add max-version here ?

**Review by DeepDiver1975 — DISMISSED** [review]
There is yet another location in the code which checks for min and max - OC_App afaik

File: lib/private/App/CodeChecker/InfoChecker.php [hunk 2]
```diff
@@ -94,17 +100,19 @@ public function analyse($appId) {
 			];
 		}
 
-		if (isset($info['dependencies']['owncloud']['@attributes']['min-version']) && ($info['requiremin'] || $info['require'])) {
+		$hasDeprecatedMin = isset($info['requiremin']) || isset($info['require']);
```
[the remark this answers is not in the document]
Comment (VicDeo) [reply lib/private/App/CodeChecker/InfoChecker.php]: with pleasure :)

**Comment by VicDeo** [comment]
@PVince81 @DeepDiver1975 another review please

File: lib/private/App/DependencyAnalyzer.php [hunk 3]
```diff
@@ -295,27 +295,25 @@ private function analyzeOC(array $dependencies, array $appInfo) {
...
 		}
 
 		if ($minVersion !== null) {
 			if ($this->compareSmaller($this->platform->getOcVersion(), $minVersion)) {
 				$missing[] = (string)$this->l->t('ownCloud %s or higher is required.', $minVersion);
 			}
+		} else {
+			$missing[] = (string)$this->l->t('missing minimal ownCloud version in appinfo.xml.');
 		}
-		if ($maxVersion !== null && !\in_array($this->platform->getOcChannel(), ['git', 'daily'], true)) {
```
Comment (DeepDiver1975) [inline lib/private/App/DependencyAnalyzer.php]: please keep this check for the channel - we need this for development and CI purpose

**Review by DeepDiver1975 — DISMISSED** [review]

File: lib/private/App/DependencyAnalyzer.php (hunk 3 above)
Comment (VicDeo) [reply lib/private/App/DependencyAnalyzer.php]: mhm. Then I need to bypass **all** checks, not only max-version as before.

Comment (DeepDiver1975) [reply lib/private/App/DependencyAnalyzer.php]: if min/max are not set at all: :bomb: 
if git/daily channel: max version criteria is allowed not to be fulfilled

Comment (VicDeo) [reply lib/private/App/DependencyAnalyzer.php]: ok, this part is moved to the next condition

File: lib/private/App/DependencyAnalyzer.php [hunk 4]
```diff
@@ -295,27 +295,27 @@ private function analyzeOC(array $dependencies, array $appInfo) {
...
-		} elseif (isset($appInfo['requiremax'])) {
-			$maxVersion = $appInfo['requiremax'];
 		}
 
 		if ($minVersion !== null) {
 			if ($this->compareSmaller($this->platform->getOcVersion(), $minVersion)) {
 				$missing[] = (string)$this->l->t('ownCloud %s or higher is required.', $minVersion);
 			}
+		} else {
+			$missing[] = (string)$this->l->t('missing minimal ownCloud version in appinfo.xml.');
```
Comment (DeepDiver1975) [inline lib/private/App/DependencyAnalyzer.php]: Better message: No minimal ownCloud version is defined in appinfo.xml.

@settermjd @phil-davis for native speaker review - THX

File: lib/private/App/DependencyAnalyzer.php [hunk 5]
```diff
@@ -295,27 +295,27 @@ private function analyzeOC(array $dependencies, array $appInfo) {
...
-		if ($maxVersion !== null && !\in_array($this->platform->getOcChannel(), ['git', 'daily'], true)) {
-			if ($this->compareBigger($this->platform->getOcVersion(), $maxVersion)) {
+		if ($maxVersion !== null) {
+			if (!\in_array($this->platform->getOcChannel(), ['git', 'daily'], true)
+				&& $this->compareBigger($this->platform->getOcVersion(), $maxVersion)
+			) {
 				$missing[] = (string)$this->l->t('ownCloud %s or lower is required.', $maxVersion);
 			}
+		} else {
+			$missing[] = (string)$this->l->t('missing maximal ownCloud version in appinfo.xml.');
```
Comment (DeepDiver1975) [inline lib/private/App/DependencyAnalyzer.php]: see above ....

**Review by DeepDiver1975 — DISMISSED** [review]

File: lib/private/App/DependencyAnalyzer.php (hunk 4 above)
Comment (phil-davis) [reply lib/private/App/DependencyAnalyzer.php]:
```
No minimum ownCloud version is defined in appinfo/info.xml
```
@settermjd ?

File: lib/private/App/DependencyAnalyzer.php (hunk 5 above)
Comment (phil-davis) [reply lib/private/App/DependencyAnalyzer.php]:
```
No maximum ownCloud version is defined in appinfo/info.xml
```
@settermjd ?

File: lib/private/App/CodeChecker/InfoChecker.php (hunk 1 above)
Comment (VicDeo) [reply lib/private/App/CodeChecker/InfoChecker.php]: the approach has changed, the attributes are checked directly as they are not on the top level.

**Review by PVince81 — APPROVED** [review]
Looks good to me :+1:

**Comment by PVince81** [comment]
no backport, this is something we said would be adressed in OC 11

**Comment by phil-davis** [comment]
> no backport, this is something we said would be adressed in OC 11

@PVince81 still true? or should any or all of this now be in the "backport to stable10" list?

**Pull request merged**